libs/wingsdk/src/std/test-runner.ts

```diff
@@ -40,18 +40,51 @@ export class TestRunner extends Resource {
       // Node.of(root).root
       Node._markRoot(Root);
 
+      let roots: IConstruct[] = [];
       if (app.isTestEnvironment) {
-        new Root(app, "env0");
+        let firstRoot = new Root(app, "env0");
+        roots.push(firstRoot);
         const tests = app._testRunner!.findTests();
         for (let i = 1; i < tests.length; i++) {
-          new Root(app, "env" + i);
+          roots.push(new Root(app, "env" + i));
+        }
+
+        // Remove test constructs that don't belong in the current environment.
+        // this should leave a single test in each environment.
+        for (let i = 0; i < tests.length; i++) {
+          let currentTestSuffix = tests[i].node.path.substring(
+            firstRoot.node.path.length
+          );
+          TestRunner._disconnectTree(roots[i], (node) => {
+            return (
+              node instanceof Test &&
+              !node.node.path.endsWith(currentTestSuffix)
+            );
+          });
         }
       } else {
         new Root(app, "Default");
       }
     }
   }
 
+  /**
+   * Scan the given construct tree and disconnect nodes that don't match the predicate.
+   * @internal
+   */
+  public static _disconnectTree(
```

Comment:
This feels wrong... I think there's already logic in the TestRunner that conditionally adds the `cloud.Function` only in the relevant environment. Perhaps there's a bug there?

See: https://github.com/winglang/wing/blob/c800787c2dac74a32079426fcfddf3acb058dc1c/libs/wingsdk/src/std/test-runner.ts#L118

Reply:
At any rate, I would highly recommend against removing nodes from the construct tree. It's a very flaky business. Instead we just need to make sure that the `cloud.Function` of the test will only be created in the test environment it belongs.

An alternative approach to the `_addTestFunction()` we have is to expose a public method on `std.Test` called something like `enableTest()` and call it on the relevant test for each enviornment.

Reply:
> This feels wrong... I think there's already logic in the TestRunner that conditionally adds the cloud.Function only in the relevant environment.

You're right, kind of surprising implementation. Anyway I'm scrapping this. :man_shrugging: